packages/next-auth/src/nextAuthPlugin.ts

```diff
@@ -3,39 +3,71 @@ import {
   PartialStlContext,
   MakeStainlessPlugin,
   Params,
+  StlContext,
 } from "stainless";
 import { IncomingMessage, ServerResponse } from "http";
 import { NextServerContext } from "@stl-api/next";
 import { AuthOptions, Session, getServerSession } from "next-auth";
 
+/// An authenticated session, with user data available.
+export type UserSession = Session & {
+  user: {
+    name?: string | null
+    email?: string | null
+    image?: string | null
+  }
+}
+
 declare module "stainless" {
   interface StlContext<EC extends AnyEndpoint> {
-    session?: Session | null;
+    /// If a route is authenticated, handlers will always have access to a valid user session.
+    /// Otherwise, a session is provided if a user is logged in.
+    session: EC['config'] extends { authenticated: true } ? UserSession : UserSession | undefined;
+  }
+
+  interface EndpointConfig {
+    authenticated?: boolean,
   }
 }
 
+function isUserSession(session: Session): session is UserSession {
+  return !!session.user;
+}
+
 export const makeNextAuthPlugin =
   ({ authOptions }: { authOptions: AuthOptions }): MakeStainlessPlugin<any> =>
-  (stl) => ({
-    async middleware<EC extends AnyEndpoint>(
-      endpoint: EC,
-      params: Params,
-      context: PartialStlContext<any, EC>
-    ) {
-      const {
-        args: [req, res],
-      } = requireNextServerContext(context);
-      // TODO catch invalid credentials errors,
-      // send appropriate error response,
-      // and somehow signal to stl.execute to early
-      // exit
-      if (req instanceof IncomingMessage && res instanceof ServerResponse) {
-        context.session = await getServerSession(req, res, authOptions);
-      } else {
-        context.session = await getServerSession(authOptions);
-      }
-    },
-  });
+    (stl) => ({
+      async middleware<EC extends AnyEndpoint>(
+        endpoint: EC,
+        params: Params,
+        context: PartialStlContext<any, EC>
+      ) {
+        const {
+          args: [req, res],
+        } = requireNextServerContext(context);
+
+        // TODO catch invalid credentials errors,
+        // send appropriate error response,
+        // and somehow signal to stl.execute to early
+        // exit
+
+        let session: Session | null | undefined;
+
+        if (req instanceof IncomingMessage && res instanceof ServerResponse) {
+          session = await getServerSession(req, res, authOptions);
+        } else {
+          session = await getServerSession(authOptions);
+        }
+        
+        if (session && isUserSession(session)) context.session = session;
+        
+        // If the endpoint requires authentication, but no user is logged in,
+        // throw unauthorized
+        if (endpoint?.config?.authenticated) {
+            if (!context.session) throw new stl.UnauthorizedError();
```

Comment:
Is there any context we could add here to return a better error message to the user?

Reply:
I don't think so, no, but what we could potentially do is provide a config option for the user to provide additional data on lack of authentication, which we'd pass to `UnauthorizedError`.

Reply:
ok, we can keep that in mind for later!